WcaOnRails/app/controllers/registrations_controller.rb

```diff
@@ -377,83 +377,121 @@ def register
+          confirmation_method: "manual",
+          confirm: true,
+          description: "Registration payment for #{competition.name} by #{registration.user.name}",
+          metadata: registration_metadata,
+        }
+        # Log the payment attempt
+        stripe_charge = StripeCharge.create!(
+          metadata: payment_intent_args.to_json,
+          stripe_charge_id: nil,
+          status: "unknown",
+        )
+        # Create the PaymentIntent, overriding the stripe_account for the request
+        # by the connected stripe account for the competition.
+        intent = Stripe::PaymentIntent.create(
+          payment_intent_args,
+          stripe_account: registration.competition.connected_stripe_account_id,
+        )
+      elsif params[:payment_intent_id]
+        stripe_charge = StripeCharge.find_by(stripe_charge_id: params[:payment_intent_id])
+        # We should definitely find a StripeCharge for this PI, so show an error if we don't.
+        unless stripe_charge
+          render json: { error: { message: t("registrations.payment_form.errors.intent_not_found") } }
+          return
+        end
+        intent = Stripe::PaymentIntent.confirm(
+          params[:payment_intent_id],
+          {},
+          stripe_account: registration.competition.connected_stripe_account_id,
+        )
+      end
+    rescue Stripe::CardError => e
+      # Log and display error to client
+      stripe_charge.update!(status: "failure", error: e.message)
+      render json: { error: { message: e.message } }
+      return
+    end
+    status, response = generate_payment_response(registration, intent, stripe_charge)
+    render json: response, status: status
   end
 
-  private def journaled_stripe_charge(*stripe_charge_create_args)
-    # Talk to Stripe to make a charge, but "journal" the attempt.
-    #  1. Before talking to Stripe, we first create a StripeCharge in status "unknown".
-    #  2. Talk to Stripe.
-    #  3. After talking to Stripe, we update that StripeCharge to status to "success" or "failure".
-    # If anything else happens (maybe our server crashed during step 2, or
-    # between step 2 and 3), then we'll have a StripeCharge in our database
-    # with "unknown" status, and we'll know to investigate it, most likely by
-    # visiting our Stripe dashboard to see if the charge actually happened or not.
-
-    stripe_charge = StripeCharge.create!(
-      metadata: stripe_charge_create_args.to_json,
-      stripe_charge_id: nil,
-      status: "unknown",
-    )
-
-    begin
-      charge = Stripe::Charge.create(*stripe_charge_create_args)
-      stripe_charge.update!(
-        status: "success",
-        stripe_charge_id: charge.id,
-      )
-    rescue Stripe::CardError, Stripe::InvalidRequestError => e
+  private def generate_payment_response(registration, intent, stripe_charge)
+    if intent&.status == "requires_action" &&
+       intent.next_action.type == "use_stripe_sdk"
```

Comment:
I went for `intent &&`, because using `&.` on the second part is useless (because as you noticed a `nil` intent will never pass the left hand side of the `&&`).